server/cmwell-bg/src/main/scala/cmwell/bg/ImpStream.scala

```diff
@@ -821,39 +825,39 @@ class ImpStream(partition: Int, config: Config, irwService: IRWService, zStore:
             ).map { bulkIndexResult =>
               // filter out expected es failures (due to we're at least once and not exactly once)
               val filteredResults = bulkIndexResult match {
-                case s@SuccessfulBulkIndexResult(_, failed) if failed.size > 0 =>
+                case s@SuccessfulBulkIndexResult(_, failed) if failed.nonEmpty =>
                   s.copy(failed = failed.filterNot { f =>
-                    (f.reason.startsWith("DocumentAlreadyExistsException") || f.reason.startsWith("VersionConflictEngineException"))
+                    f.reason.startsWith("DocumentAlreadyExistsException") ||
+                    f.reason.startsWith("VersionConflictEngineException")
                   })
                 case other => other
               }
-              BGMessage(bgMessages.map(_.offsets).flatten, (filteredResults, bgMessages.map {
-                _.message._2
-              }.flatten))
+              BGMessage(bgMessages.flatMap(_.offsets), (filteredResults, bgMessages.flatMap(_.message._2)))
             }
         }
       )
 
       val manageIndices = builder.add(
-        Flow[BGMessage[(BulkIndexResult, Seq[IndexCommand])]].mapAsync(1) {
-          case bgMessage@BGMessage(_, (bulkRes, commands)) =>
-            startingIndexCount += bulkRes.successful.size
+        Flow[((String,Long),BGMessage[(BulkIndexResult, Seq[IndexCommand])])].mapAsync(1) {
+          case (unchangedState@(currentIndexName,indexCount),bgMessage@BGMessage(_, (bulkRes, commands))) =>
+            val numOfShardPerIndex = ftsService.numOfShardsForIndex(currentIndexName)
+            var startingIndexCount = indexCount + bulkRes.successful.size
             if (startingIndexCount / numOfShardPerIndex >= maxDocsPerShard) {
               val (pref, suf) = currentIndexName.splitAt(currentIndexName.lastIndexOf('_') + 1)
               val nextCount = suf.toInt + 1
               val nextIndexName = pref + nextCount
               cmwell.util.concurrent.retry(3)(ftsService.createIndex(nextIndexName)).flatMap { _ =>
                 scheduleFuture(10.seconds)(ftsService.updateAllAlias)
               }.map { case _ =>
-                currentIndexName = nextIndexName
-                startingIndexCount = 0
-                bgMessage
+                ((nextIndexName,0L),bgMessage)
               }
             } else
-              Future.successful(bgMessage)
+              Future.successful(unchangedState -> bgMessage)
```

Comment:
The new count is never passed. Should be `(currentIndexName, startingIndexCount)`.